### PR TITLE
Handle read only apps folders properly

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
This is planned to support clustered setup much better ....

## Related Issue
refs  https://github.com/owncloud/platform/issues/133

## How Has This Been Tested?
1. add the config parameter 'cluster.mode' -> true
2. try to install
3. try to operate as usual
4. expectation: nothing will break apart ....

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: lib/private/Config.php [hunk 1]
```diff
@@ -272,5 +281,12 @@ private function writeData() {
 			\OC_Util::clearOpcodeCache();
 		}
 	}
+
+	public function isReadOnly() {
+		if (!$this->getValue('installed', false)) {
```
Comment (PVince81) [inline lib/private/Config.php]: also include the `config_is_read_only` key here ?

File: lib/private/Installer.php [hunk 2]
```diff
@@ -91,7 +91,11 @@ public static function installApp( $data = []) {
 
 		$info = self::checkAppsIntegrity($data, $extractDir, $path);
 		$appId = OC_App::cleanAppId($info['id']);
-		$basedir = OC_App::getInstallPath().'/'.$appId;
+		$appsFolder = OC_App::getInstallPath();
+		if ($appsFolder === null) {
```
Comment (PVince81) [inline lib/private/Installer.php]: looks like you should also add `is_writeable` and `is_readable` here ? See https://github.com/owncloud/core/pull/28543/files#diff-ec8c121d705ff47eacbfdd23f36a9f6cL731.

File: settings/templates/panels/admin/mail.php [hunk 3]
```diff
@@ -22,87 +22,94 @@
...
-					<?php if ($_['mail_smtpmode'] != 'smtp') print_unescaped(' class="hidden"'); ?>>
-				<?php foreach ($mail_smtpsecure as $secure => $name):
-					$selected = '';
-					if ($secure == $_['mail_smtpsecure']):
-						$selected = 'selected="selected"';
-					endif; ?>
-					<option value='<?php p($secure)?>' <?php p($selected) ?>><?php p($name) ?></option>
-				<?php endforeach;?>
-			</select>
+			<?php p($l->t('The config file is read only. Please adopt your setup by editing the config file manually.')); ?>
```
Comment (PVince81) [inline settings/templates/panels/admin/mail.php]: adopting a child ?

=> adjust

File: settings/templates/panels/admin/mail.php [hunk 4]
```diff
@@ -22,87 +22,94 @@
...
+			<br>
+			<?php p($l->t('In a clustered setup please make sure to sync the file across all nodes.')); ?>
```
Comment (PVince81) [inline settings/templates/panels/admin/mail.php]: s/the file/the config.php file/

to make sure this is clear

**Review by PVince81 — COMMENTED** [review]
Please add back the check for "config_is_read_only" as upgrading to 10.0.3 would likely cause surprises for people who have it set.

If we want to remove it then deprecate it in 10.0.3 and remove completely in 11.0. But I'm not sure if we should remove it as some people without cluster setup might want to have the config read-only, so they wouldn't think about setting cluster mode.

**Comment by DeepDiver1975** [comment]
@PVince81 thanks for your feedback - I removed the config parameter 'cluster.mode' now totally. 'config_is_read_only' will do its job for now

**Comment by PVince81** [comment]
hmm, so does it mean that "config_read_only" will now also block the apps folder ? People who had this set will be surprised that they can't install new apps ?

**Comment by DeepDiver1975** [comment]
> hmm, so does it mean that "config_read_only" will now also block the apps folder ? People who had this set will be surprised that they can't install new apps ?

Thats my point about this all: it's a mess and needs cleanup .... this was the idea in having ONE config switch to rule them all ... 😉

**Comment by PVince81** [comment]
@DeepDiver1975 do we want to surprise these people ? If yes, will need proper communication

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 do we want to surprise these people ? If yes, will need proper communication

so you advocate for keeping two intendant checks
- config_is_read_only for config stuff only
- and test if apps folder is writable

**Comment by PVince81** [comment]
If this was OC 11 I'd be fine with breaking the way it works.

For OC 10 the easiest is just do nothing and maybe add a note about clustering in the documentation about the two settings to change:
- config_is_read_only
- the "apps_paths" setting with "writable" set to false

Now if the above doesn't work correctly then we should fix that at least.

**Comment by DeepDiver1975** [comment]
@PVince81 a much simpler solution now - only addressing if the apps folder is writable

**Review by PVince81 — APPROVED** [review]
:+1: that's better, thanks.

Please rename issue title and merge if you tested this and all still worked fine

**Pull request merged**